File: src/main/org/nlogo/swing/TextField.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.swing;

// thank you thank you thank you to
//   http://cvs.sourceforge.net/cgi-bin/viewcvs.cgi/uic/uicompiler/uic/widgets/UICTextField.java

public strictfp class TextField
    extends javax.swing.JTextField {

  private boolean mouseEvent = false;

  public TextField(int columns) {
    super(columns);
    setDragEnabled(false);
    initListener();
  }

  public TextField(String text, int columns) {
    super(text, columns);
    setDragEnabled(false);
    initListener();
  }

  public TextField(javax.swing.text.Document doc, String text, int columns) {
    super(doc, text, columns);
    setDragEnabled(false);
    initListener();
  }

  protected void initListener() {
    enableEvents(java.awt.AWTEvent.MOUSE_EVENT_MASK);
    addFocusListener
        (new java.awt.event.FocusListener() {
          public void focusGained(java.awt.event.FocusEvent fe) {
            if (!mouseEvent) {
              // this is like selectAll(), but it leaves the
              // caret at the beginning rather than the start;
              // this prevents the enclosing scrollpane from
              // scrolling to the end to make the caret
              // visible; it's nicer to keep the scroll at the
              // start - ST 12/20/04
              setCaretPosition(getText().length());
              moveCaretPosition(0);
            }
          }

          public void focusLost(java.awt.event.FocusEvent fe) {
            mouseEvent = fe.isTemporary();
          }
        });
  }

  @Override
  public void processMouseEvent(java.awt.event.MouseEvent me) {
    if (me.getID() == java.awt.event.MouseEvent.MOUSE_PRESSED) {
      mouseEvent = true;
    }
    super.processMouseEvent(me);
  }

}

File: netlogo-gui/src/main/swing/TextField.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.swing

import java.awt.{ Graphics, Insets }
import java.awt.event.{ FocusEvent, FocusListener, KeyEvent, MouseAdapter, MouseEvent }
import javax.swing.{ JTextField, KeyStroke }
import javax.swing.border.LineBorder
import javax.swing.text.Document

import org.nlogo.theme.{ InterfaceColors, ThemeSync }

class TextField(columns: Int = 0, text: String = null, document: Document = null)
  extends JTextField(document, text, columns) with ThemeSync {

  private lazy val undoManager = new UndoManager

  private var mouseEvent = false

  TextActions.applyToComponent(this)

  undoManager.watch(this)

  getKeymap.addActionForKeyStroke(KeyStroke.getKeyStroke(KeyEvent.VK_Z, getToolkit.getMenuShortcutKeyMaskEx),
                                  UndoManager.undoAction)

  getKeymap.addActionForKeyStroke(KeyStroke.getKeyStroke(KeyEvent.VK_Y, getToolkit.getMenuShortcutKeyMaskEx),
                                  UndoManager.redoAction)

  addMouseListener(new MouseAdapter {
    override def mousePressed(e: MouseEvent): Unit = {
      mouseEvent = true
    }
  })

  addFocusListener(new FocusListener {
    def focusGained(e: FocusEvent): Unit = {
      if (!mouseEvent) {
        // this is like selectAll(), but it leaves the
        // caret at the beginning rather than the start;
        // this prevents the enclosing scrollpane from
        // scrolling to the end to make the caret
        // visible; it's nicer to keep the scroll at the
        // start - ST 12/20/04
        setCaretPosition(getText.size)
        moveCaretPosition(0)
      }
    }

    def focusLost(e: FocusEvent): Unit = {
      mouseEvent = e.isTemporary
    }
  })

  syncTheme()

  override def setText(text: String): Unit = {
    super.setText(text)

    undoManager.discardAllEdits()
  }

  override def getInsets: Insets =
    new Insets(0, 3, 0, 0)

  override def paintComponent(g: Graphics): Unit = {
    if (isEnabled) {
      setBackground(InterfaceColors.textAreaBackground())
    } else {
      setBackground(InterfaceColors.Transparent)
    }

    super.paintComponent(g)
  }

  override def syncTheme(): Unit = {
    setForeground(InterfaceColors.textAreaText())
    setCaretColor(InterfaceColors.textAreaText())
    setBorder(new LineBorder(InterfaceColors.textAreaBorderEditable()))
  }
}
